sites/game-client/ui/context-menu.mjs

```diff
@@ -0,0 +1,157 @@
+// Copyright 2021 HITCON Online Contributors
+// SPDX-License-Identifier: BSD-2-Clause
+
+const CONTEXT_MENU_OTHER_DIV = 'context-menu-other';
+const CONTEXT_MENU_LIST_OTHER_UL = 'context-menu-list-other';
+const CONTEXT_MENU_SELF_DIV = 'context-menu-self';
+const CONTEXT_MENU_LIST_SELF_UL = 'context-menu-list-self';
+
+const mapCellSize = 32;
+
+/**
+ * Maintain the state of the context menu for players.
+*/
+class ContextMenu {
+    /**
+     *
+     */
+    constructor(gameState, mapRenderer) {
+        this.gameState = gameState;
+        this.mapRenderer = mapRenderer;
+        this.othersMenu = new Map();
+        this.selfMenu = new Map();
+
+        this.focusedPlayer = null;
+        this.playerInfo = null;
+
+        // bind the contextmenu and click event.
+        document.getElementById('mapcanvas').addEventListener('contextmenu', this.onContextMenu.bind(this));
+        document.body.addEventListener('click', this.hideMenu);
+
+        // Get current player
+        window.addEventListener('gameStart', (d) => {
+            this.playerInfo = d.detail.gameClient.playerInfo;
+        });
+    }
+
+    /**
+     * Called when user clicked on the player.
+     * @param {Event} e event
+     */
+    onContextMenu(e) {
+        e.preventDefault();
+
+        // Check if the user actually clicked on a player.
+        if (!this.findFocusedUser(e.pageX, e.pageY)) {
+            return;
+        }
+
+        // Identify whether the menu is for the user itself or other players.
+        // Stop if the menu is empty.
+        let div;
+        if (this.focusedPlayer.playerID === this.playerInfo.playerID && this.selfMenu.size > 0) {
+            div = CONTEXT_MENU_SELF_DIV;
+        } else if (this.focusedPlayer.playerID !== this.playerInfo.playerID && this.othersMenu.size > 0) {
+            div = CONTEXT_MENU_OTHER_DIV;
+        } else {
+            this.focusedPlayer = null;
+            return;
+        }
+
+        let menu = document.getElementById(div);
+        if (menu.style.display == "block") {
+            this.hideMenu();
```

Comment:
Maybe a comment on what this block is for? i.e. menu was already open.

Reply:
Also, if the menu is already open, maybe we should close it first then open the current one?

Reply:
I'm not sure what's the common design of context menu, but at least on my machine, the second right-click would close the existing context menu.